### PR TITLE
[reland] "[reland] `_foreach_copy` with different src/dst dtypes"

### LINKED ISSUE #115171: [Feature] `_foreach_copy_` supports different src/dst dtype with fusion
Today, if you use `torch._foreach_copy_(dsts, srcs)` where `dsts` are in one dtype and `srcs` are in another, the foreach copy takes the slow path and does not batch any of the copies:
https://github.com/pytorch/pytorch/blob/e06bff8bbea84ce672285bc34690b2e45a1b63ab/aten/src/ATen/native/cuda/ForeachBinaryOpList.cu#L265-L269
The feature request is for the copies to be batched so that the copying achieves better GPU bandwidth and has lower CPU overhead.

cc: @janeyx99 @crcrpar 

cc @crcrpar @mcarilli

### PR DESCRIPTION
**Pull request by crcrpar**
Fixes #115171

### CODE REVIEW

**Comment by ezyang** [comment]
So what caused the IMAs

**Comment by crcrpar** [comment]
> So what caused the IMAs

lack of array bounds checking and missing dispatch to int

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/native/cuda/ForeachBinaryOpList.cu [hunk 1]
```diff
@@ -252,20 +254,156 @@ FOREACH_BINARY_OP_LIST(
...
+  }
+};
+
+#define AT_DISPATCH_SOURCE_TYPES(TYPE, NAME, ...)                                  \
+  AT_DISPATCH_SWITCH(                                                              \
+      TYPE,                                                                        \
+      NAME,                                                                        \
+      AT_PRIVATE_CASE_TYPE_USING_HINT(                                             \
+          at::ScalarType::Byte, src_t, __VA_ARGS__)                                \
+          AT_PRIVATE_CASE_TYPE_USING_HINT(                                         \
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/ForeachBinaryOpList.cu:294]: turn off clang format here lol

Comment (crcrpar) [reply aten/src/ATen/native/cuda/ForeachBinaryOpList.cu:294]: Glad you didn't love this format haha

**Pull request closed without merging**